src/appleseed.studio/mainwindow/rendering/lightpathswidget.cpp

```diff
@@ -160,149 +193,539 @@ void LightPathsWidget::slot_synchronize_camera()
-    static const GLfloat LightDiffuse[] = { 0.8f, 0.8f, 0.8f, 1.0f };
-    static const GLfloat LightSpecular[] = { 0.3f, 0.3f, 0.3f, 1.0f };
+        if (!success)
+        {
+            char info_log[1024];
+            glGetShaderInfoLog(shader, 1024, NULL, info_log);
 
-    glEnable(GL_LIGHT0);
-    glLightfv(GL_LIGHT0, GL_POSITION, LightPosition);
-    glLightfv(GL_LIGHT0, GL_AMBIENT, LightAmbient);
-    glLightfv(GL_LIGHT0, GL_DIFFUSE, LightDiffuse);
-    glLightfv(GL_LIGHT0, GL_SPECULAR, LightSpecular);
-}
+            GLint shader_kind;
+            glGetShaderiv(shader, GL_SHADER_TYPE, &shader_kind);
+            string shader_kind_string = shader_kind_to_string(shader_kind);
 
-void LightPathsWidget::resizeGL(int w, int h)
-{
-    glViewport(0, 0, static_cast<GLsizei>(w), static_cast<GLsizei>(h));
-}
+            RENDERER_LOG_ERROR("opengl: %s shader compilation failed:\n%s", shader_kind_string.c_str(), info_log);
+        }
+    }
 
-namespace
-{
-    void glMultMatrixd(const Matrix4d& m)
+    void link_shader_program(
+        const GLuint program,
+        const GLuint vert,
+        const GLuint frag)
     {
-        const Matrix4d mt(transpose(m));
-        ::glMultMatrixd(const_cast<GLdouble*>(&mt[0]));
+        glAttachShader(program, vert);
+        glAttachShader(program, frag);
+        glLinkProgram(program);
+
+        GLint success;
+        glGetProgramiv(program, GL_LINK_STATUS, &success);
+
+        if (!success)
+        {
+            char info_log[1024];
+            glGetProgramInfoLog(program, 1024, NULL, info_log);
+            RENDERER_LOG_ERROR("opengl: shader program linking failed:\n%s", info_log);
+        }
     }
 
-    void glColor(const Color3f& c)
+    void create_shader_program(
+        GLuint&                 program,
+        const string&      vert_source,
+        const string&      frag_source)
     {
-        glColor3f(c[0], c[1], c[2]);
+
+        GLuint vert = glCreateShader(GL_VERTEX_SHADER);
+        GLuint frag = glCreateShader(GL_FRAGMENT_SHADER);
+
+        auto gl_vert_source = static_cast<const GLchar*>(vert_source.c_str());
+        auto gl_frag_source = static_cast<const GLchar*>(frag_source.c_str());
+
+        compile_shader(vert, 1, &gl_vert_source, NULL);
+        compile_shader(frag, 1, &gl_frag_source, NULL);
+
+        program = glCreateProgram();
+        link_shader_program(program, vert, frag);
+
+        glDeleteShader(vert);
+        glDeleteShader(frag);
     }
 
+
     struct OpenGLRasterizer
-      : public ObjectRasterizer
+        : public ObjectRasterizer
     {
-        void begin_object() override
-        {
-            glBegin(GL_TRIANGLES);
-        }
+        vector<float> buffer;
+        size_t             prim_count;
 
-        void end_object() override
+        void begin_object(const size_t triangle_count_hint) override
         {
-            glEnd();
+            buffer.clear();
+            buffer.reserve(triangle_count_hint * 18);
```

Comment:
Constant 18 is used in multiple places, make it a real constant and document its meaning (3 vertices, each with 3D position and normal).